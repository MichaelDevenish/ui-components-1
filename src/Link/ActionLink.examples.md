#### Basic action link

```jsx
<ActionLink
  href='#'
>
Your text here
</ActionLink>
```

#### Action link with alternate anchor tag

```jsx
const { DefaultsProvider } = require("../Defaults");

<DefaultsProvider
  value={{
    AnchorComponent: 'i'
  }}
>
  <ActionLink
    href='#'
  >
    Your text here
  </ActionLink>
</DefaultsProvider>
```
