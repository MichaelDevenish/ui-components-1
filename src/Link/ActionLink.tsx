import React from 'react'
import styled from 'styled-components'
import { AnchorProps, withAnchor } from '../Anchor'

const ActionLinkLabel = styled.span`
  color: rgb(5, 87, 223);
  font-weight: 600;
  font-size: .9375rem;
  text-transform: uppercase;
`

export interface ActionLinkProps extends AnchorProps {}
export const ActionLink: React.ComponentClass<ActionLinkProps> = withAnchor(ActionLinkLabel)
