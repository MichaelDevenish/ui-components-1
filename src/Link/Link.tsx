import React from 'react'
import styled from 'styled-components'
import { AnchorProps, withAnchor } from '../Anchor'

const TextLinkLabel = styled.span`
  color: rgb(5, 87, 223);
`

const ActionLinkLabel = styled.span`
  color: rgb(5, 87, 223);
  font-weight: 600;
  font-size: .9375rem;
  text-transform: uppercase;
`

export interface TextLinkProps extends AnchorProps {}
export const TextLink: React.ComponentClass<TextLinkProps> = withAnchor(TextLinkLabel)

export interface ActionLinkProps extends AnchorProps {}
export const ActionLink: React.ComponentClass<ActionLinkProps> = withAnchor(ActionLinkLabel)
