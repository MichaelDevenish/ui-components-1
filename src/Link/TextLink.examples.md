#### Basic text link

```jsx
<TextLink
  href='#'
>
Your text here
</TextLink>
```

