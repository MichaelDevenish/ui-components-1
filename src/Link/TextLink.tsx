import React from 'react'
import styled from 'styled-components'
import { AnchorProps, withAnchor } from '../Anchor'

const TextLinkLabel = styled.span`
  color: rgb(5, 87, 223);
`

export interface TextLinkProps extends AnchorProps {}
export const TextLink: React.ComponentClass<TextLinkProps> = withAnchor(TextLinkLabel)
