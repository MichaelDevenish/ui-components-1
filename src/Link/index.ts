export { ActionLink } from './ActionLink'
export { TextLink } from './TextLink'
