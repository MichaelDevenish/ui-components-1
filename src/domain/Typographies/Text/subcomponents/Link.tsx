import React, { useContext } from 'react'

import { isExternalURL } from '../../../Internals/Anchor/helper'
import { StyledA } from './style'

interface ILinkProps {
  /** Href to link to */
  href?: string
  /* Callback to fire on click */
  onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void
  /** @deprecated Use `openInNewTab` instead. Href target */
  target?: string
  /** Open href in new tab */
  openInNewTab?: boolean
  /** Underlines the text when hovered */
  underlineOnHover?: boolean
  /** Specifies the style variant */
  variant?: LinkVariant
}

enum LinkVariant {
  Default = 'default',
  Unstyled = 'unstyled'
}

const Link: React.FC<ILinkProps> = ({ children, href, onClick, variant, underlineOnHover, target, openInNewTab }) => {
  const targetProp = openInNewTab ? '_blank' : target
  let relProp: string | undefined

  if (href && isExternalURL(href)) {
    const rels = new Set((relProp ?? '').split(/\s+/).filter(Boolean as any))
    rels.add('noreferrer')

    relProp = Array.from(rels.values()).join(' ')
  }

  return (
    <StyledA
      href={href}
      onClick={onClick}
      target={targetProp}
      rel={relProp}
      role={(!!onClick && !href) ? 'button' : undefined}
      variant={variant || LinkVariant.Default}
      underlineOnHover={underlineOnHover}
    >
      {children}
    </StyledA>
  )
}

export {
  Link,
  ILinkProps,
  LinkVariant
}
