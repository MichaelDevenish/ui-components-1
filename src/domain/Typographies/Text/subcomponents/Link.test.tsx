import { mount } from 'enzyme'
import React from 'react'

import { Link } from './Link'

describe('<Link />', () => {
  it('should render a link', () => {
    const wrapper = mount(
      <>
        <Link href='#'>Relative URL</Link>
        <Link href='http://localhost/#'>Absolute URL</Link>
      </>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a link with rel when href is external', () => {
    const wrapper = mount(
      <Link href='http://www.google.com'>Hello World</Link>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a link with target when openInNewTab is true', () => {
    const wrapper = mount(
      <Link href='#' openInNewTab>Hello World</Link>
    )

    expect(wrapper).toMatchSnapshot()
  })
})
