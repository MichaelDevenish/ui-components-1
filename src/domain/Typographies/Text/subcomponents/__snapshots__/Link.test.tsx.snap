// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should render a link 1`] = `
Array [
  .c0 {
  -webkit-transition: color .25s ease-out;
  transition: color .25s ease-out;
  cursor: pointer;
}

.c0,
.c0:link,
.c0:visited {
  color: #432DF3;
}

.c0:hover {
  color: #2512B3;
}

.c0:active {
  color: #0F0080;
}

<Link
    href="#"
  >
    <styled.a
      href="#"
      variant="default"
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "c0",
              "rules": Array [
                "
  &:hover {
    ",
                [Function],
                "
  }

  ",
                [Function],
                "
",
              ],
            },
            "displayName": "styled.a",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "a",
            "toString": [Function],
            "usesTheme": false,
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        href="#"
        variant="default"
      >
        <a
          className="c0"
          href="#"
        >
          Relative URL
        </a>
      </StyledComponent>
    </styled.a>
  </Link>,
  .c0 {
  -webkit-transition: color .25s ease-out;
  transition: color .25s ease-out;
  cursor: pointer;
}

.c0,
.c0:link,
.c0:visited {
  color: #432DF3;
}

.c0:hover {
  color: #2512B3;
}

.c0:active {
  color: #0F0080;
}

<Link
    href="http://localhost/#"
  >
    <styled.a
      href="http://localhost/#"
      variant="default"
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "c0",
              "rules": Array [
                "
  &:hover {
    ",
                [Function],
                "
  }

  ",
                [Function],
                "
",
              ],
            },
            "displayName": "styled.a",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "a",
            "toString": [Function],
            "usesTheme": false,
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        href="http://localhost/#"
        variant="default"
      >
        <a
          className="c0"
          href="http://localhost/#"
        >
          Absolute URL
        </a>
      </StyledComponent>
    </styled.a>
  </Link>,
]
`;

exports[`<Link /> should render a link with rel when href is external 1`] = `
.c0 {
  -webkit-transition: color .25s ease-out;
  transition: color .25s ease-out;
  cursor: pointer;
}

.c0,
.c0:link,
.c0:visited {
  color: #432DF3;
}

.c0:hover {
  color: #2512B3;
}

.c0:active {
  color: #0F0080;
}

<Link
  href="http://www.google.com"
>
  <styled.a
    href="http://www.google.com"
    rel="noreferrer"
    variant="default"
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  &:hover {
    ",
              [Function],
              "
  }

  ",
              [Function],
              "
",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "a",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      href="http://www.google.com"
      rel="noreferrer"
      variant="default"
    >
      <a
        className="c0"
        href="http://www.google.com"
        rel="noreferrer"
      >
        Hello World
      </a>
    </StyledComponent>
  </styled.a>
</Link>
`;

exports[`<Link /> should render a link with target when openInNewTab is true 1`] = `
.c0 {
  -webkit-transition: color .25s ease-out;
  transition: color .25s ease-out;
  cursor: pointer;
}

.c0,
.c0:link,
.c0:visited {
  color: #432DF3;
}

.c0:hover {
  color: #2512B3;
}

.c0:active {
  color: #0F0080;
}

<Link
  href="#"
  openInNewTab={true}
>
  <styled.a
    href="#"
    target="_blank"
    variant="default"
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  &:hover {
    ",
              [Function],
              "
  }

  ",
              [Function],
              "
",
            ],
          },
          "displayName": "styled.a",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "a",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      href="#"
      target="_blank"
      variant="default"
    >
      <a
        className="c0"
        href="#"
        target="_blank"
      >
        Hello World
      </a>
    </StyledComponent>
  </styled.a>
</Link>
`;
